crates/red_knot_python_semantic/src/types/class.rs

```diff
@@ -938,14 +949,65 @@ impl<'db> KnownClass {
     }
 
     pub(crate) fn to_instance(self, db: &'db dyn Db) -> Type<'db> {
-        self.to_class_literal(db).to_instance(db)
+        self.to_class_literal(db)
+            .into_class_literal()
+            .map(|ClassLiteralType { class }| Type::instance(class))
+            .unwrap_or_else(Type::unknown)
+    }
+
+    pub(crate) fn try_to_class_literal(
+        self,
+        db: &'db dyn Db,
+    ) -> Result<ClassLiteralType<'db>, KnownClassLookupError<'db>> {
+        let symbol = known_module_symbol(db, self.canonical_module(db), self.as_str(db)).symbol;
+        match symbol {
+            Symbol::Type(Type::ClassLiteral(class_type), Boundness::Bound) => Ok(class_type),
+            Symbol::Type(Type::ClassLiteral(class_type), Boundness::PossiblyUnbound) => {
+                Err(KnownClassLookupError::ClassPossiblyUnbound { class_type })
+            }
+            Symbol::Type(found_type, _) => {
+                Err(KnownClassLookupError::SymbolNotAClass { found_type })
+            }
+            Symbol::Unbound => Err(KnownClassLookupError::ClassNotFound),
+        }
     }
 
     pub(crate) fn to_class_literal(self, db: &'db dyn Db) -> Type<'db> {
-        known_module_symbol(db, self.canonical_module(db), self.as_str(db))
-            .symbol
-            .ignore_possibly_unbound()
-            .unwrap_or(Type::unknown())
+        // a cache of the `KnownClass`es that we have already failed to lookup in typeshed
+        // (and therefore that we've already logged a warning for)
+        static MESSAGES: LazyLock<Mutex<FxHashSet<KnownClass>>> = LazyLock::new(Mutex::default);
+
+        self.try_to_class_literal(db)
+            .map(Type::ClassLiteral)
+            .unwrap_or_else(|lookup_error| {
+                assert!(
+                    !cfg!(any(test, debug_assertions)),
+                    "{}",
+                    lookup_error.display(db, self)
+                );
```

Comment:
I changed this so that it panics if _either_ the `test` feature is enabled _or_ the `debug_assertions` feature is enabled. That's because we have some tests for which the `test` feature is apparently not enabled (mdtests), and we have some tests that we only ever run in release-mode in CI (property tests). I think we probably want both to panic if a known-class symbol lookup unexpectedly fails, rather than silently falling back to `Unknown`.

Reply:
The downside of panicking in tests is that we now never actually test the fallback behavior... So we won't know if it works as expected. I'm inclined to remove the panic case.